### ISSUE TITLE
Accept header-based negotiation

### ISSUE THREAD

**Issue by nerdstein** [d8-contrib-modules/services#4]
Research and implement

**Comment by nerdstein** [comment]
@kylebrowning - need some use cases here to understand context

**Comment by kylebrowning** [comment]
http://www.metaltoad.com/blog/why-drupal-8-wont-ship-with-REST-content-negotiation

**Comment by nerdstein** [comment]
@kylebrowning - how does this relate to issue #2? Is this just related to data formats?

**Comment by nerdstein** [comment]
I'm going to spend some time reviewing http://symfony.com/doc/current/cmf/components/routing/nested_matcher.html to see if we can override or extend Drupal's OOTB routing to extend content negotiation

**Comment by nerdstein** [comment]
Pre-execute event subscriber may be the way to go for content negotiation:

http://symfony.com/doc/current/cookbook/event_dispatcher/before_after_filters.html

**Comment by nerdstein** [comment]
This is adopted in Drupal and can likely address issues around authentication and other content negotiation needs

**Comment by nerdstein** [comment]
This is the "validation" step, or filter, before executing a request. It can also throw exceptions to block

**Comment by nerdstein** [comment]
We can check the "Accept" value from the request and match this up with a data format.

**Comment by nerdstein** [comment]
"Accept-Language" should also be discussed due to the D8 enhancements

**Comment by nerdstein** [comment]
TL;DR:

Content negotiation can be done with an event subscriber and handle the validation of request. This will impact 1. authentication 2. data formats 3. localization. Need to research "Route Filters" further, especially in D8 (http://symfony.com/doc/current/cmf/components/routing/nested_matcher.html#the-route-filters) to see what role it can play.

**Comment by nerdstein** [comment]
Closing, new tickets coming
